core/trino-main/src/test/java/io/trino/sql/planner/TestLogicalPlanner.java

```diff
@@ -1707,6 +1707,33 @@ public void testRedundantLimitNodeRemoval()
                         values(ImmutableList.of("x"))));
     }
 
+    @Test
+    public void testLimitPushdownThroughUnionNesting()
+    {
+        String query = "SELECT col FROM (" +
+                "SELECT nationkey FROM nation " +
+                "UNION ALL " +
+                "SELECT nationkey FROM nation " +
+                "UNION ALL " +
+                "SELECT nationkey FROM nation) AS t(col) LIMIT 2";
+        assertPlan(
+                query,
+                output(
+                        limit(
+                                2,
+                                ImmutableList.of(),
+                                false,
+                                exchange(
+                                        LOCAL,
+                                        GATHER,
+                                        exchange(
```

Comment:
why there is repartition before gather? Seems wrong

Reply:
That's what the logic in `AddLocalExchanges` does. It will insert a local repartitioning exchange to achieve thread-level parallelism. It's not necessary for this particular plan shape, but that's been the long standing behavior even before these changes.